src/System.ServiceModel.Syndication/src/System/ServiceModel/Syndication/XmlReaderWrapper.cs

```diff
@@ -13,311 +13,86 @@ internal class XmlReaderWrapper : XmlReader
-        private Func<XmlReaderWrapper, Task> _skipFunc;
-        private Func<XmlReaderWrapper, Task<bool>> _readFunc;
-        private Func<XmlReaderWrapper, Task<string>> _readInnerXmlFunc;
-
-        private XmlReaderWrapper(XmlReader reader)
+        public static XmlReader CreateFromReader(XmlReader reader)
         {
-            if (reader == null)
-            {
-                throw new ArgumentNullException(nameof(reader));
-            }
-
-            _reader = reader;
-
-            if (_reader.Settings.Async)
-            {
-                InitAsync();
-            }
-            else
+            if (reader is XmlReaderWrapper || reader.Settings.Async)
             {
-                Init();
+                return reader;
             }
-        }
 
-        public static XmlReaderWrapper CreateFromReader(XmlReader reader)
-        {
-            XmlReaderWrapper wrappedReader = reader as XmlReaderWrapper;
-            return wrappedReader != null ? wrappedReader : new XmlReaderWrapper(reader);
+            return new XmlReaderWrapper(reader);
         }
 
-        private void InitAsync()
+        private XmlReaderWrapper(XmlReader reader)
         {
-            _getValueFunc = new Func<XmlReaderWrapper, Task<string>>((thisPtr) => { return thisPtr._reader.GetValueAsync(); });
-            _moveToContentFunc = new Func<XmlReaderWrapper, Task<XmlNodeType>>((thisPtr) => { return thisPtr._reader.MoveToContentAsync(); });
-            _skipFunc = new Func<XmlReaderWrapper, Task>((thisPtr) => { return thisPtr._reader.SkipAsync(); });
-            _readFunc = new Func<XmlReaderWrapper, Task<bool>>((thisPtr) => { return thisPtr._reader.ReadAsync(); });
-            _readInnerXmlFunc = new Func<XmlReaderWrapper, Task<string>>((thisPtr) => { return thisPtr._reader.ReadInnerXmlAsync(); });
-        }
+            _reader = reader ?? throw new ArgumentNullException(nameof(reader));
 
-        private void Init()
-        {
-            _getValueFunc = new Func<XmlReaderWrapper, Task<string>>((thisPtr) => { return Task.FromResult<string>(thisPtr._reader.Value); });
-            _moveToContentFunc = new Func<XmlReaderWrapper, Task<XmlNodeType>>((thisPtr) => { return Task.FromResult<XmlNodeType>(_reader.MoveToContent()); });
-            _skipFunc = new Func<XmlReaderWrapper, Task>((thisPtr) => { _reader.Skip(); return Task.CompletedTask; });
-            _readFunc = new Func<XmlReaderWrapper, Task<bool>>((thisPtr) => { return Task.FromResult<bool>(_reader.Read()); });
-            _readInnerXmlFunc = new Func<XmlReaderWrapper, Task<string>>((thisPtr) => { return Task.FromResult<string>(_reader.ReadInnerXml()); });
-        }
-
-        public override XmlNodeType NodeType
-        {
-            get
+            if (_reader.Settings.Async)
             {
-                return _reader.NodeType;
+                throw new ArgumentException("Async XmlReader should not be wrapped", nameof(reader));
```

Comment:
Should we add a resx string for the message?

Reply:
No meed, the only way this can happen is with a bug in the code as the type is internal and the only caller should by the static wrapping method.